### PR TITLE
[WIP/testing] Multi jvm test reporting

### PR DESCRIPTION
**Pull request by jrudolph**
I finally had a successful stab at trying out how to improve test reporting for multi-jvm tests. Test reporting currently don't works properly because sbt-multi-jvm built its own infrastructure for forking test launchers (maybe because it existed before forking tests was implemented in sbt). When forking tests it's tricky to collect test events live at runtime because those need to be communicated from the forked process to sbt in a structured way.

For testing, I moved the sbt-multi-jvm sources over into the akka sources and adapted the task for running multi-jvm tests locally to use the internal sbt forking to get full test report coverage.

I open the PR here only to let it run through the PR validator to see if jenkins can now pick up the test results.

### CODE REVIEW

File: project/plugins.sbt [hunk 1]
```diff
@@ -1,6 +1,11 @@
 // these comment markers are for including code into the docs
 //#sbt-multi-jvm
-addSbtPlugin("com.typesafe.sbt" % "sbt-multi-jvm" % "0.4.0")
+//addSbtPlugin("com.typesafe.sbt" % "sbt-multi-jvm" % "0.4.0")
```
Comment (helena) [inline project/plugins.sbt]: this a ToDo remove lines 2 and 3 if all goes well?

File: project/src/main/scala/com/typesafe/sbt/SbtMultiJvm.scala [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+
+import com.typesafe.sbt.multijvm.{Jvm, JvmLogger}
+import com.typesafe.sbt.multijvm.Compat.{Process, _}
+import com.typesafe.sbt.multijvm.Compat.Implicits._
+import sbt._
+import Keys._
+import java.io.File
+import java.lang.Boolean.getBoolean
+
+import scala.Console.{GREEN, RESET}
```
Comment (helena) [inline project/src/main/scala/com/typesafe/sbt/SbtMultiJvm.scala]: Nice

**Review by helena — COMMENTED** [review]
Looking forward to trying this

**Comment by patriknw** [comment]
Probably not related to this PR but in same area; suddenly `akka-multi-node-nightly` is failing with this message at the end:
```
ERROR: Step ‘Publish JUnit test result report’ failed: Test reports were found but none of them are new. Did leafNodes run? 
For example, /localhome/jenkinsakka/workspace/akka-multi-node-nightly/akka-http-core/target/test-reports/TEST-akka.http.impl.engine.client.ClientCancellationSpec.xml is 2 yr 6 mo old
```

I can't see in the job description where akka-http-core is defined.

**Comment by jrudolph** [comment]
Yeah, sorry, that's because multi-jvm tests don't report test results (yet!) and I disabled the other tests for that job yesterday.

**Comment by jrudolph** [comment]
I actually checked the "Don't fail the build if test results are empty" box in the configuration to prevent the build from failing, but it seems it now fails because it only finds old builds. I now also added cleaning the working with git before the build, so maybe that fixes it.

File: project/plugins.sbt [hunk 3]
```diff
@@ -1,6 +1,11 @@
...
+libraryDependencies += Defaults.sbtPluginExtra(
+  "com.eed3si9n" % "sbt-assembly" % "0.14.5",
+  (sbtBinaryVersion in pluginCrossBuild).value,
+  (scalaBinaryVersion in pluginCrossBuild).value
+)
 //#sbt-multi-jvm
```
Comment (patriknw) [inline project/plugins.sbt]: note that this is inside a doc snippet, is this what we want to show in docs?

**Comment by patriknw** [comment]
so the intention is to contribute this back upstream when it has been verified?

**Comment by jrudolph** [comment]
> so the intention is to contribute this back upstream when it has been verified?

Yes, running it with PR validation didn't make any sense though, as PR validation isn't running the multi-jvm tests...

File: project/plugins.sbt (hunk 3 above)
Comment (jrudolph) [reply project/plugins.sbt]: That's just for testing it right inside the akka build.

**Review by helena — APPROVED** [review]
I'd like to see this in action, LGTM aside from the docs question

**Comment by raboof** [comment]
What is the status of this work? Did we test this branch with `akka-multi-node-nightly` yet?

**Comment by jrudolph** [comment]
It doesn't work for multi node at all. It only reports for multi-jvm tests so far.

**Comment by jrudolph** [comment]
To make it work with multi-node, one has to manage to run the tests in forked mode on a remote node and be able connect to the sbt reporting channel as it would be done locally for forked tests. I guess it's possible but it's not as easy to do.

**Comment by jrudolph** [comment]
Closing for now.

**Pull request closed without merging**